File: loadmorefooter/src/main/java/com/takwolf/android/hfrecyclerview/loadmorefooter/LoadMoreState.java
package com.takwolf.android.hfrecyclerview.loadmorefooter;

import androidx.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@IntDef({LoadMoreState.DISABLED, LoadMoreState.LOADING, LoadMoreState.FINISHED, LoadMoreState.ENDLESS, LoadMoreState.FAILED})
@Retention(RetentionPolicy.SOURCE)
public @interface LoadMoreState {
    int DISABLED = 0;
    int LOADING = 1;
    int FINISHED = 2;
    int ENDLESS = 3;
    int FAILED = 4;
}

File: paging/src/main/java/com/takwolf/android/hfrecyclerview/paging/LoadMoreState.kt
package com.takwolf.android.hfrecyclerview.paging

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

enum class LoadMoreState(val canLoadMore: Boolean) {
    DISABLED(false),
    IDLE(true),
    LOADING(false),
    FINISHED(false),
    FAILED(true),
}

fun StateFlow<LoadMoreState>.observe(
    owner: LifecycleOwner,
    loadMoreFooter: LoadMoreFooter,
) {
    owner.lifecycleScope.launch {
        owner.repeatOnLifecycle(Lifecycle.State.STARTED) {
            collect { state ->
                loadMoreFooter.state = state
            }
        }
    }
}
